frontend/rust-lib/flowy-sqlite/src/search.rs

```diff
@@ -0,0 +1,298 @@
+use diesel::{sql_query, sql_types::Text, QueryResult, RunQueryDsl, SqliteConnection};
+
+const SEARCH_INDEX_TABLE: &str = "search_index";
+const CREATE_SEARCH_INDEX_TABLE: &str =
+  "CREATE VIRTUAL TABLE if not exists search_index USING fts5(index_type, view_id, id, data)";
+
+/// Runs database migrations for local search using sqlite FTS5.
+///
+/// FTS5 tables do not have indexes, which is not supported by Diesel.
+pub fn run_migrations(conn: &mut SqliteConnection) -> QueryResult<usize> {
+  if !table_exists(conn, SEARCH_INDEX_TABLE)? {
+    sql_query(CREATE_SEARCH_INDEX_TABLE).execute(conn)?;
+  }
+  Ok(0)
+}
+
+#[derive(Debug, QueryableByName)]
+struct ShowTablesRow {
+  #[allow(dead_code)]
+  #[diesel(sql_type=Text)]
+  name: String,
+}
+
+fn table_exists(conn: &mut SqliteConnection, table: &str) -> QueryResult<bool> {
+  let tables: Vec<ShowTablesRow> =
+    sql_query("SELECT name FROM sqlite_master WHERE type='table' AND name=?")
+      .bind::<Text, _>(table)
+      .load(conn)?;
+  Ok(!tables.is_empty())
+}
+
+/// The type of data that is stored in the search index row.
+#[derive(Debug)]
+pub enum IndexType {
+  /// Name of the view is stored in data.
+  View,
+  /// Text of the document is stored in data.
+  Document,
+}
+
+impl ToString for IndexType {
+  fn to_string(&self) -> String {
+    match self {
+      IndexType::View => "view".to_owned(),
+      IndexType::Document => "document".to_owned(),
+    }
+  }
+}
+
+/// SearchData represents the data that is contained by a view and the type of document.
+#[derive(Debug, PartialEq, QueryableByName)]
+pub struct SearchData {
+  /// The type of data that is stored in the search index row.
+  #[diesel(sql_type = Text)]
+  pub index_type: String,
+
+  /// The `View` that the row references.
+  #[diesel(sql_type = Text)]
+  pub view_id: String,
+
+  /// The ID that corresponds to the type that is stored.
+  /// View: view_id
+  /// Document: page_id
+  #[diesel(sql_type = Text)]
+  pub id: String,
+
+  /// The data that is stored in the search index row.
+  #[diesel(sql_type = Text)]
+  pub data: String,
+}
+
+impl SearchData {
+  pub fn new_view(view_id: &str, name: &str) -> Self {
+    Self {
+      index_type: IndexType::View.to_string(),
+      view_id: view_id.to_owned(),
+      id: view_id.to_owned(),
+      data: name.to_owned(),
+    }
+  }
+
+  pub fn new_document(view_id: &str, page_id: &str, text: &str) -> Self {
+    Self {
+      index_type: IndexType::Document.to_string(),
+      view_id: view_id.to_owned(),
+      id: page_id.to_owned(),
+      data: text.to_owned(),
+    }
+  }
+}
+
+/// Add search data for searching.
+pub fn add(conn: &mut SqliteConnection, data: &SearchData) -> QueryResult<usize> {
```

Comment:
Can we use the tokio::task::spawn_blocking if there are too many add called

Reply:
Do you mean wrap the calls to add in `spawn_blocking`? Currently, the runs in a task spawned by `af_spawn` in `folder_indexer.rs`, so the calls can be wrapped there.